lib/Dialect/ATen/ATenLoweringPass.cpp

```diff
@@ -0,0 +1,972 @@
+    Value rhs = MemRefTypeCast(rewriter, operands[1]);
+    auto indexType = IndexType::get(op->getContext());
+
+    using namespace edsc;
+
+    ScopedContext scope(rewriter, loc);
+    Value zero = intrinsics::std_constant_index(0);
+    Value one = intrinsics::std_constant_index(1);
+    MemRefBoundsCapture vRes(result), vLHS(lhs), vRHS(rhs);
+    StdIndexedValue iRes(result), iLHS(lhs), iRHS(rhs);
+    Value M(vRes.ub(0));
+    if (vRes.rank() == 1) {
+      affineLoopNestBuilder({zero}, {M}, 1, [&](ValueRange ivs) {
+        Value i = ivs[0];
+        iRes(i) = iLHS(i) + iRHS(i);
+      });
+    } else if (vRes.rank() == 2) {
+      Value N(vRes.ub(1));
+      affineLoopNestBuilder({zero, zero}, {M, N}, {1, 1}, [&](ValueRange ivs) {
+        Value i = ivs[0];
+        Value j = ivs[1];
+        iRes(i, j) = iLHS(i, j) + iRHS(i, j);
+      });
+    } else if (vRes.rank() == 3) {
+      Value N(vRes.ub(1));
+      Value O(vRes.ub(2));
+      affineLoopNestBuilder({zero, zero, zero}, {M, N, O}, {1, 1, 1},
+                            [&](ValueRange ivs) {
+                              Value i = ivs[0];
+                              Value j = ivs[1];
+                              Value k = ivs[2];
+                              iRes(i, j, k) = iLHS(i, j, k) + iRHS(i, j, k);
+                            });
+    } else {
+      Value N(vRes.ub(1));
+      Value O(vRes.ub(2));
+      Value P(vRes.ub(3));
+      affineLoopNestBuilder({zero, zero, zero, zero}, {M, N, O, P},
+                            {1, 1, 1, 1}, [&](ValueRange ivs) {
+                              Value i = ivs[0];
+                              Value j = ivs[1];
+                              Value k = ivs[2];
+                              Value l = ivs[3];
+                              iRes(i, j, k, l) =
+                                  iLHS(i, j, k, l) + iRHS(i, j, k, l);
+                            });
+    }
+    // Return the newly allocated buffer.
+    rewriter.replaceOp(op, {result});
+    return success();
+  }
+};
+
+
+// Replace the given operation with a call to the given function.
+// The function is assumed to accept memrefs and scalar types and return
+// Memrefs. Here the result types are converted back to the result types of op,
+// but operands are NOT converted.  This allows non-standard mappings from
+// operand types to function types.
+LogicalResult
+rewriteWithVoidFunctionCallExplicit(Operation *op,
+                                    ArrayRef<Value> callops,
+                                    ArrayRef<Value> operands,
+                                    ConversionPatternRewriter &rewriter,
+                                    std::string functionName) {
+
+    auto loc = op->getLoc();
+    edsc::ScopedContext scope(rewriter, loc);
+
+    // The original operation types.
+    SmallVector<Type, 8> opTys;
+    // Shape erased versions of the original operation types.
+    SmallVector<Type, 8> erasedOpTys;
+    for (auto &o: callops) {
```

Comment:
```suggestion
    for (Value &o: callops) {
```

(avoid `auto` when it does not improve readability)